File: Leetcode/Problems C++/162. Find Peak Element.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int beg = 0;
        int end = nums.size();
        if(end<=1)return 0;
        end-=1;
        
        if(nums[beg]>nums[beg+1])return beg;
        if(nums[end]>nums[end-1])return end;
        while(beg<=end)
        {
            int mid = beg + (end-beg)/2;
            int next = mid+1;
            int prev = mid-1;
            if(nums[mid]>nums[next] && nums[mid]>nums[prev])
                return mid;
            else if(nums[mid]<nums[next])
            {
                beg=mid+1;
            }
            else
            {
                end = mid-1;
            }
        }
        return -1;
        
    }
};

File: Leetcode/Revision/Binary Search/162. Find Peak Element.java
class Solution {
    public int findPeakElement(int[] nums) {
        int n = nums.length;
        if(n==1)return 0;
        if(nums[0]>nums[1])return 0;
        if(nums[n-1]>nums[n-2])return n-1;
        int s = 1;
        int e = n-2;
        
        while(s<=e){
            int mid = s+(e-s)/2; 
            if(nums[mid-1]<nums[mid] && nums[mid+1]<nums[mid]){
                return mid;
            }
            else if(nums[mid-1]>nums[mid]){
                e = mid-1;
            }
            else{
                s=mid+1;
            }
        }
        return -1;
    }
}
